File: JavaScript/0349-两个数组的交集.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
function intersection(nums1, nums2) {
  if (nums1.length < nums2.length) {
    [nums1, nums2] = [nums2, nums1];
  }
  const set = new Set(nums1),
    result = new Set();

  for (let i = nums2.length - 1; i >= 0; i--) {
    const value = nums2[i];
    set.has(value) && result.add(value);
  }

  return Array.from(result);
}

File: Java/0349-两个数组的交集.java
class Solution {
  public int[] intersection(int[] nums1, int[] nums2) {
    int[] hash = new int[1002];
    int[] result = new int[1002];
    int n = 0;
    for (int i = 0, length = nums1.length; i < length; i++) {
      hash[nums1[i]] = 1;
    }
    for (int i = 0, length = nums2.length; i < length; i++) {
      int num = nums2[i];
      if (hash[num] == 1) {
        hash[num]++;
        result[n] = num;
        n++;
      }
    }
    return Arrays.copyOfRange(result, 0, n);
  }
}
